src/mdio/converters/segy_to_mdio_v1.py

```diff
@@ -0,0 +1,241 @@
+    segy_spec: SegySpec,
+    mdio_template: AbstractDatasetTemplate,
+    overwrite: bool = False,
+):
+    """A function that converts a SEG-Y file to an MDIO v1 file.
+    """
+    # Open a SEG-Y file according to the SegySpec
+    # This could be a spec from the registryy or a custom spec
+    segy_settings = SegySettings(storage_options=input.options)
+    segy_file = SegyFile(url=input.uri, spec=segy_spec, settings=segy_settings)
+
+    # Extract dimensions, coordinates, and index_headers
+    # This is an expensive read operation performed in chunks
+    dimensions, coordinates, index_headers = _scan_for_dims_coords_headers(
+        segy_file, mdio_template)
+    shape = [len(dim.coords) for dim in dimensions]
+
+    # Specify the header structure for the MDIO dataset
+    # TODO: Setting headers (and thus creating a variable with StructuredType data type)
+    # causes the error in _populate_dims_coords_and_write_to_zarr originating
+    # dask/array/core.py:1346 
+    #   TypeError: Cannot cast array data from dtype([('trace_seq_num_line', '<i4'), ... 
+    #   to dtype('bool') according to the rule 'unsafe'
+    # I believe Dask does not support StructuredType data type.
+    # Thus, we are not setting headers for now:
+    # headers = None
+    headers = to_structured_type(index_headers.dtype)
+    # TODO: Set Units to None for now, will fix this later
+    mdio_ds: Dataset = mdio_template.build_dataset(name="NONE", 
+                                                   sizes=shape, 
+                                                   coord_units=None,
+                                                   headers=headers)
+    xr_dataset: xr_Dataset = to_xarray_dataset(mdio_ds=mdio_ds)
+
+    # Validate SEG-Y dimensions and headers against xarray dataset created from the template.
+    _validate_segy(coordinates=coordinates, templated_dataset=xr_dataset)
+
+    # Write the xr Dataset with coords and dimensions, but empty traces and headers.
+    # Drop them after the write, so we can write the data variable in chunks later
+    # Writing traces and headers is an memory-expensive and time-consuming operation
+    # We will write them in chunks later
+    xr_dataset = _populate_dims_coords_and_write_to_zarr(dataset=xr_dataset,
+                                                         dimensions=dimensions,
+                                                         coordinates=coordinates,
+                                                         output=output)
+
+    grid_map = _populate_trace_mask_and_write_to_zarr(segy_file=segy_file,
+                                                      dimensions=dimensions,
+                                                      index_headers=index_headers,
+                                                      xr_sd=xr_dataset,
+                                                      output=output)
+
+    # TODO: Maybe we should have saved the data_variable_name as a Dataset attribute?
```

Comment:
yea we should do that, we need to update the schema if we want to enforce it. DatasetMetadata can have a field for "default_variable" that stores this value.